cs3api4lab/api/cs3apismanager.py

```diff
@@ -422,15 +416,29 @@ def _save_directory(self, path):
 
     @asyncify
     def _check_write_permissions(self, path):
-
         parent = self._get_parent_path(path)
-        stat = self.file_api.stat_info(parent)
-        if not ShareUtils.map_permissions_to_role(stat['permissions']) == 'editor':
+        file_info = self.file_api.stat_info(parent)
+
+        if not self._is_editor(file_info):
             raise web.HTTPError(403, u'The path %s is not writable' % parent)
-        # check if the path is a received share with editor permissions
-        for share in self.share_api.list_received()['content']:
-            if share['path'] == path and share['writable'] == False:
-                raise web.HTTPError(403, u'The share %s is not writable' % path)
+
+    def _is_editor(self, file_info):
+        is_editor = False
+
+        if not file_info:
+            return is_editor
+
+        # check file permissions
+        if ShareUtils.map_permissions_to_role(file_info['permissions']) == 'editor':
+            is_editor = True
+
+        # check if file is shared with me
+        role = self.share_api.get_share_received_role(file_info['filepath'])
```

Comment:
I do not understand why do we need to check the share permission. I should not have file permissions and then lose them...

Reply:
Right now because of the content manager validation, there is no way to determine if we are opening a file from a file browser or a shared file. There is no way of passing other parameters that are not specified in endpoint validation, so everything is handled by one function. The issue is, that if the file is opened/saved plugin doesn't know if it's a shared file or a file from the user's account. This is why we are checking permissions for files and for a received share, according to API, these permissions are separate entities so they both need to be checked.

Reply:
document... All of these implementation details should always be documented in the code...

Reply:
@glpatcern @gmgigi96 is it true that file permissions and share permissions are independent? And, should I care? I might be accessing a share, but if I get permission via some other path, I believe the most permissive acls should win, no?

Reply:
@piotrWichlinskiSoftwaremind @diocas the short answer is you should not care. The backend (reva) is expected to sort out the permissions from the access token, no matter which way (share, public link, etc) the user got that token.

Now we know this is not entirely true in some cases, but it will be fixed. Therefore, in such API I'd suggest to just check the file's permissions.